File: pauli/Pauli.py
from __future__ import annotations

import numpy as np
import F2_helper.F2_helper as f2
import numba

class Pauli: # TODO add from_matrix method
    # Pauli is (-1)^sign bit * (-i) *=^ (i_bit) * X^x_vector * Z^z_vector
    def __init__(self, number_qubits : int, x_vector : int, z_vector : int, sign_bit : int, i_bit : int):
        self.number_qubits = number_qubits
        self.x_vector = x_vector
        self.z_vector = z_vector
        
        self.sign_bit = sign_bit
        self.i_bit = i_bit

        self.update_phase()

    def update_phase(self):
        self.phase = (1-2*self.sign_bit)*(1+(-1j-1)*self.i_bit)

    def generate_matrix(self) -> np.ndarray:
        size = 2**self.number_qubits
        matrix = np.zeros((size, size), dtype=complex)

        for j in range(size):
            matrix[j^self.x_vector, j] = self.phase*f2.sign_mod2product(j, self.z_vector)

        return matrix
    
    def multiply_by_pauli_on_right(self, other_pauli : Pauli):
        self.sign_bit ^= other_pauli.sign_bit ^ f2.mod2product(self.z_vector, other_pauli.x_vector) ^ (self.i_bit & other_pauli.i_bit)
        self.i_bit ^= other_pauli.i_bit

        self.x_vector ^= other_pauli.x_vector
        self.z_vector ^= other_pauli.z_vector

        self.update_phase()

    def is_hermitian(self) -> bool:
        return self.i_bit == f2.mod2product(self.x_vector, self.z_vector)
    
    def anticommutes_with(self, other_pauli : Pauli) -> int:
        return f2.mod2product(self.x_vector, other_pauli.z_vector) ^ f2.mod2product(self.z_vector, other_pauli.x_vector)
    
    def commutes_with(self, other_pauli : Pauli) -> int:
        return 1 ^ self.anticommutes_with(other_pauli)
    
    def check_sign_eigenstate(self, state_vector : np.ndarray, num_qubits : int, sign_bit : int) -> bool: # TODO
        phase = (1-2*sign_bit) * self.phase
        
        for index in range(1 << num_qubits):
            if state_vector[index ^ self.x_vector] != phase * f2.sign_mod2product(index, self.z_vector) * state_vector[index]:
                return False

        return True
    
    def __eq__(self, other : object) -> bool:
        if not isinstance(other, Pauli):
            return False
        
        result = True
        result &= (self.sign_bit == other.sign_bit)
        result &= (self.i_bit == other.i_bit)
        result &= (self.number_qubits == other.number_qubits)
        result &= (self.x_vector == other.x_vector)
        result &= (self.z_vector == other.z_vector)
        result &= (self.phase == other.phase)

        return result

@numba.njit()
def round_to_5dp(value : complex) -> complex:
    return round(value.real, 5) + 1j*round(value.imag, 5)

File: src/pauli/pauli.cpp
#include "pauli.h"
#include "f2_helper.h"

namespace fst
{
    Pauli::Pauli(const std::size_t number_qubits, const std::size_t x_vector, const std::size_t z_vector, const bool sign_bit, const bool imag_bit)
    : number_qubits(number_qubits)
    , x_vector(x_vector)
    , z_vector(z_vector)
    , sign_bit(sign_bit)
    , imag_bit(imag_bit)
    {
        update_phase();
    }

    void Pauli::update_phase()
    {
        phase = { f_min1_pow( sign_bit ) * float_not( imag_bit ), f_min1_pow( sign_bit ) * static_cast<float>( imag_bit ) };
    }

    bool Pauli::is_hermitian() const
    {
        return imag_bit == f2_dot_product(x_vector, z_vector);
    }

    bool Pauli::anticommutes_with(const Pauli &other_pauli) const
    {
        return f2_dot_product(x_vector, other_pauli.z_vector) ^ f2_dot_product(z_vector, other_pauli.x_vector);
    }

    bool Pauli::commutes_with(const Pauli &other_pauli) const
    {
        return !anticommutes_with(other_pauli);
    }

    std::vector<std::vector<std::complex<float>>> Pauli::get_matrix() const 
    {
        const std::size_t size = integral_pow_2( number_qubits );
        std::vector<std::vector<std::complex<float>>> matrix (size, std::vector<std::complex<float>> (size, 0));

        for ( size_t col_index = 0; col_index < size; col_index++ )
        {
            matrix.at( col_index ^ x_vector).at(col_index) = 
                phase * sign_f2_dot_product( col_index, z_vector );
        }
        
        return matrix;
    }

    std::vector<std::complex<float>> Pauli::multiply_vector (const std::vector<std::complex<float>> &vector) const
    {
        const size_t size = vector.size();
        std::vector<std::complex<float>> result(size, 0);

        for( size_t index = 0; index < size; index++)
        {
            result[index^x_vector] = phase * sign_f2_dot_product(index, z_vector) ;
        }

        return result;
    }

    void Pauli::multiply_by_pauli_on_right(const Pauli &other_pauli)
    {
        imag_bit ^= other_pauli.imag_bit;
        int sign_bit_update = f2_dot_product(z_vector, other_pauli.x_vector) ^ (imag_bit & other_pauli.imag_bit);
        sign_bit ^= sign_bit_update;
        update_phase();

        x_vector ^= other_pauli.x_vector;
        z_vector ^= other_pauli.z_vector;
    }

    bool Pauli::has_eigenstate(const std::vector<std::complex<float>> &vector, const unsigned int eig_sign) const
    {
        const std::size_t size = integral_pow_2(number_qubits);
        const std::complex<float> vector_phase = f_min1_pow( eig_sign ) * phase;

        for ( size_t index = 0; index < size; index++)
        {
            std::complex<float> expected_phase = vector_phase * sign_f2_dot_product(index, z_vector) * vector.at(index);
            if (vector.at(index ^ x_vector) != expected_phase) {
                return false;
            }
        }

        return true;
    }

}